src/state.rs

```diff
@@ -158,21 +165,22 @@ fn animate(
             &mut TextureAtlasSprite,
             &Handle<SpriteSheetAnimation>,
             &mut SpriteSheetAnimationState,
+            Option<&PlaySpeedMultiplier>,
         ),
         With<Play>,
     >,
 ) {
-    for (entity, sprite, animation, mut state) in
-        animations
-            .iter_mut()
-            .filter_map(|(entity, sprite, anim_handle, state)| {
+    for (entity, sprite, animation, mut state, optional_speed_multiplier) in
+        animations.iter_mut().filter_map(
+            |(entity, sprite, anim_handle, state, optional_speed_multiplier)| {
                 animation_defs
                     .get(anim_handle)
                     .filter(|anim| anim.has_frames())
-                    .map(|anim| (entity, sprite, anim, state))
-            })
+                    .map(|anim| (entity, sprite, anim, state, optional_speed_multiplier))
+            },
+        )
     {
-        if state.update(sprite, animation, time.delta()) {
+        if state.update(sprite, animation, time.delta(), optional_speed_multiplier) {
```

Comment:
We could do
```suggestion
        let speed_factor = optional_speed_multiplier.map(|m| m.0).unwrap_or(1.0);
        if state.update(sprite, animation, time.delta() * speed_factor) {
```

And then, the `update` function would not have to change at all.
My motivation is not that much that I don't want to change it. But more that I think `update` doesn't really need the complexity of an extra argument, as `delta` already contains the speed information.

The only tradeoff I can see is that it would require to write an "integration test" that knows about bevy instead of a simpler unit-test of the `update` function. But I think, it is probably worth it. I may write that test myself if you don't know how to or don't have time for it.

Reply:
I'd appreciate if you could help me on that regard, I'm still learning about the codebase and stuff. thank you!

I totally agree that that change is better since we don't have to manually pepper/update several function call sites

Reply:
`time.delta() * speed_factor` does not work directly since we don't have a Mul<f64> impl block. Trying to create one.

Reply:
code-formatting lint suggested an even nicer form, `.map_or`

```rs
        let adjusted_delta = optional_speed_multiplier.map_or(time.delta(), |m| {
            Duration::from_secs_f64(time.delta().as_secs_f64() * m.0)
        });
```